[JiaXinEu] iP

JiaXinEu opened the pull request:
# JavAssist

> “Today is the day you’re going to slay your to-do list and bring home a major win.” — Michael Hyatt ([source](https://medium.com/@oaklinejournal/25-quotes-from-successful-people-on-the-power-of-writing-to-do-lists-8a0b3729a34c#:~:text=25%20Quotes%20from%20Successful%20People%20on%20the%20power,%E2%80%9CThe%20Six%20Most%20Important%20Things.%E2%80%9D%20...%20More%20items))

JavAssist is the right tool for you. It's: 

- text-based
- easy to learn
- ~~FAST~~ _SUPER FAST_ to use

All you need to do is,

1. download it from [here](https://github.com/JiaXinEu/ip/releases/download/v0.1/duke.jar).
2. run it.
3. add your tasks.
4. let it manage your task for you 😉 

And it is **FREE**!

Features:

- [x] Managing tasks
- [x] Managing deadlines
- [x] Managing events
- [ ] Reminders (coming soon)

---

It is written in Java. Here's the `main` method:
```java
    public static void main(String[] args) {
        String logo = "\t  ____   __  _  _  __    ___  ___  ____  ___  ____\n\t"
                + " (_  _) /__\\( \\/ )/__\\  / __)/ __)(_  _)/ __)(_  _)\n\t"
                + " .-_)( /(__)\\\\  //(__)\\ \\__ \\\\__ \\ _)(_ \\__ \\  )(\n\t"
                + "\\____)(__)(__)\\/(__)(__)(___/(___/(____)(___/ (__)\n";

        new Duke("Duke.txt", "./data", "JavAssist", logo).run();
    }
```

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

Commit:
docs/README.md: Tweak document template

Let's tweak the docs/README.md (which is used as the user guide)
to fit Duke better. Specifically,

1. mention product name in the title
2. mention adding a product screenshot and a product intro
3. tweak the flow to describe feature-by-feature

JiaXinEu committed:
Rename, greet, exit

JiaXinEu committed:
Add feature to echo command and exit when input bye

JiaXinEu committed:
Add list to store and display text

JiaXinEu committed:
Add class Task to support mark task as done or not done

JiaXinEu committed:
Add subclasses Todo, Deadline, Event to support 3 types of tasks

JiaXinEu committed:
Add input.txt and EXPECTED.txt for semi-automated testing

JiaXinEu committed:
Add error handling mechanisms to Duke.java

JiaXinEu committed:
Add feature to delete task

JiaXinEu committed:
Add enum to handle input type

JiaXinEu committed:
Update Duke.java to save to hard disk

JiaXinEu committed:
Merge branch 'branch-Level-7'

JiaXinEu committed:
Modify runtest.bat

JiaXinEu committed:
Modify type of datetime in Deadline.java and Event.java

JiaXinEu committed:
Merge branch 'branch-Level-8'

JiaXinEu committed:
Add classes and DukeException.java

JiaXinEu committed:
Move classes into packages

JiaXinEu committed:
Merge branch 'branch-A-Packages'

JiaXinEu committed:
Merge branch 'add-gradle-support'

JiaXinEu committed:
Modify mainClass in build.gradle

JiaXinEu committed:
Add test folder for JUnit testing

JiaXinEu committed:
Merge branch 'branch-A-JUnit'

JiaXinEu committed:
Update details of shadowJar

JiaXinEu committed:
Merge branch 'branch-A-Jar'

JiaXinEu committed:
Add JavaDoc comments to classes and methods

JiaXinEu committed:
Modify code to follow coding standard

JiaXinEu committed:
Add search task by keyword feature

JiaXinEu committed:
Merge branch 'branch-A-JavaDoc'

JiaXinEu committed:
Merge branch 'branch-A-CodingStandard'

JiaXinEu committed:
Merge branch 'branch-Level-9'

JiaXinEu committed:
Set up checkStyle

JiaXinEu committed:
Merge branch 'branch-A-CheckStyle'

JiaXinEu committed:
Add GUI to chatbot

JiaXinEu committed:
Merge branch 'branch-Level-10'

JiaXinEu committed:
Modify find command to use Java varargs

JiaXinEu committed:
Merge branch 'branch-A-Varargs'

h4ow3i commented on src/main/java/duke/command/Command.java:
```diff
@@ -0,0 +1,15 @@
+public interface Command {
+    public String execute(TaskList list, Ui ui, Storage storage) throws DukeException;
+
+    public boolean isExit();
```
small issue but could consider renaming it to 'hasExit'.

yadunut replied:
One way to make the code cleaner is to have a default impl for isExit, 
```java
    public boolean isExit() {
        return false;
    }
```

Since most commands don't exit by default, and we can override this in the ExitCommand to return true

h4ow3i commented on src/main/java/duke/command/MarkCommand.java:
```diff
@@ -0,0 +1,68 @@
+public class MarkCommand implements Command {
+
+    private String input;
+    private boolean toMark;
```
could consider renaming it to "isMarked"

h4ow3i commented on src/main/java/duke/task/Deadline.java:
```diff
@@ -0,0 +1,68 @@
+ * Represents a deadline object.
+ */
+public class Deadline extends Task {
+    private LocalDateTime date;
```
could consider renaming this to 'due' or 'by' as the variable could be a date with time.

h4ow3i reviewed (commented):
Great code! It was clear and neat and leaves little to no room for confusion for readers.

martinng01 commented on src/main/java/duke/task/Task.java:
```diff
@@ -0,0 +1,51 @@
+    /**
+     * Marks isDone as true.
+     */
+    public void done() {
```
could consider renaming done() to markAsDone() so that it is clearer that it is setting done as true

martinng01 commented on src/main/java/duke/task/Task.java:
```diff
@@ -0,0 +1,51 @@
+    /**
+     * Marks isDone as false.
+     */
+    public void undo() {
```
could consider renaming undo() to markAsUndone() so that it is the opposite of the done() function

martinng01 commented on src/main/java/duke/command/Command.java:
I like how you abstracted command as a separate class. Very intuitive!

martinng01 commented on src/main/java/duke/command/DeadlineCommand.java:
```diff
@@ -0,0 +1,54 @@
+    }
+
+    @Override
+    public boolean equals(Object a) {
```
I like how you overrode the function equals(), makes it easy to check if two commands are the equal.

martinng01 reviewed (commented):
Overall good job with the code, very clean and follows coding standards!

JiaXinEu committed:
Add asserts to Command classes

Possible mistakes by programmer is not checked.

Check size of array before accessing elements in array.

Checking number of elements before accessing can avoid processing the
command wrongly.

Let's,
* use assert
* edit run configuration to enable assertion

JiaXinEu committed:
Modify code to ensure code quality

Code is messy and violates code quality.

Refactor code to ensure code quality.

Code quality improves the readability of code.

Let's,
* extract method
* remove magic literals
* make happy path prominent

yadunut commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,102 @@
+    private Storage storage;
+    private TaskList tasks;
+    private Ui ui;
+    private boolean isExit = false;
```
this `isExit` doesn't seem like its being used / the variable is being shadowed in `run()`. This makes it quite confusing to know whether the function-local `isExit` is being used or the object variable `isExit` is being used.

yadunut commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,102 @@
+     * @param name Name of chatbot.
+     * @param logo Logo of chatbot.
+     */
+    public Duke(String file, String fileParent, String name, String logo) {
```
Instead of taking in the `file` and the `fileParent`, you could derive the parent path from the full file path. 

For example
```java
import java.io.File;
import java.nio.file.Path;
import java.nio.file.Paths;
Path path = Paths.get(filepath);
File directory = path.getParent().toFile();
directory.mkdirs();
this.file = path.toFile();
this.file.createNewFile();
```
The above code would get the Path of the directory, create it if it hasn't been created, and then create the file if it hasn't been created also.

yadunut commented on src/main/java/duke/util/Parser.java:
```diff
@@ -0,0 +1,217 @@
+     * @param pattern Regex pattern of a valid command format.
+     * @return True if matches, false otherwise.
+     */
+    public static boolean matchPattern(String input, String pattern) {
+        Pattern regexPattern = Pattern.compile(pattern);
+        return regexPattern.matcher(input).matches();
+    }
```
Since the compiled `regexPattern` is not used, you could do
```java
boolean b = Pattern.matches("a*b", "aaaaab");
```
instead, as shown in the [docs](https://docs.oracle.com/javase/8/docs/api/java/util/regex/Pattern.html)

yadunut commented on src/main/java/duke/util/Parser.java:
```diff
@@ -0,0 +1,217 @@
+    /**
+     * Sets the possible types of input.
+     */
+    public enum InputType {
+        LIST,
+        MARK,
+        UNMARK,
+        DELETE,
+        BYE,
+        TODO,
+        DEADLINE,
+        EVENT,
+        FIND,
+        UNKNOWN
+    }
```
Nice work on using enums here :)

yadunut reviewed (commented):
Great work on the comprehensive tests and very thorough javadocs!

Some minor nits with regards to writing better/less verbose code, but otherwise great work :)

JiaXinEu committed:
Add feature to take in multiple commands

Input can only represent one command.

Handling multiple commands in on input allows series of actions on tasks.

Let's,
* use Streams to handle multiple commands

JiaXinEu committed:
Merge pull request #1 from JiaXinEu/branch-A-Assertions

Add asserts to Command classes

JiaXinEu committed:
Merge branch 'master' into branch-A-CodeQuality

JiaXinEu committed:
Merge pull request #2 from JiaXinEu/branch-A-CodeQuality

Modify code to ensure code quality

JiaXinEu committed:
Merge branch 'master' into branch-A-Streams

JiaXinEu committed:
Merge pull request #3 from JiaXinEu/branch-A-Streams

Add feature to take in multiple commands

JiaXinEu committed:
Modify code to follow checkstyle

JiaXinEu committed:
Add gradle.yml

JiaXinEu committed:
Add feature to manage expenses

JiaXinEu committed:
Merge branch 'branch-BCD-Extension'

JiaXinEu committed:
Modify code on GUI

JiaXinEu committed:
Merge branch 'branch-A-BetterGui'

JiaXinEu committed:
Modify name and personality of chatbot

JiaXinEu committed:
Merge branch 'branch-A-Personality'

JiaXinEu committed:
Update code to handle unexpected date

JiaXinEu committed:
Merge branch 'branch-A-MoreErrorHandling'

JiaXinEu committed:
Add unit tests on for new feature Expenses

JiaXinEu committed:
Merge branch 'branch-A-MoreTesting'

JiaXinEu committed:
Improve code quality with help of Ai tool

JiaXinEu committed:
Merge branch 'branch-A-AiAssisted'

JiaXinEu committed:
Update comment

JiaXinEu committed:
Upload screenshot of product

JiaXinEu committed:
Add user guide

JiaXinEu committed:
Merge branch 'branch-A-UserGuide'

JiaXinEu committed:
Update title of README.md

JiaXinEu committed:
Update README.md to correct typo

JiaXinEu committed:
Modify build.gradle to create JAR file
